File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {ListNode}
     */
    reverseList(head) {
        if(!head) return null;

        let curr = head;
        if(head.next) {
            curr = this.reverseList(head.next);
            head.next.next = head;
        }
        head.next = null;
        return curr;
    }
}

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-1.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    void insertIntoMap(Node head) {
        Node temp = head;
        while(temp != null) {
            Node nextElement = temp.next;
            Node copy = new Node(temp.val);
            copy.next = nextElement;
            temp.next = copy;
            temp = nextElement;
        }
    }
    void connectRandomPointers(Node head) {
        Node temp = head;
        while(temp != null){
            Node copy = temp.next;
            if(temp.random != null) {
                copy.random = temp.random.next;
            } else {
                copy.random = null;
            }
            temp= temp.next.next;
        }
    }
    Node getDeepCopy(Node head) {
        Node dummy = new Node(-1);
        Node temp = head;
        Node curr = dummy;

        while(temp != null) {
            curr.next = temp.next;
            curr = curr.next;

            temp.next = temp.next.next;
            temp = temp.next;
        }
        return dummy.next;
    }

    public Node copyRandomList(Node head) {
        if(head == null) return head;
        // insert into map as a copy
        insertIntoMap(head);
        // connect the random pointers
        connectRandomPointers(head);
        // create a new list from the copied and random
        return getDeepCopy(head);
    }
}
